packages/block-library/src/paragraph/use-enter.js

```diff
@@ -49,19 +57,39 @@ export function useExitOnEnterAtEnd( props ) {
 
 			const order = getBlockOrder( wrapperClientId );
 
+			event.preventDefault();
+
+			const position = order.indexOf( clientId );
 			// It should be the last block.
-			if ( order.indexOf( clientId ) !== order.length - 1 ) {
+			if ( position === order.length - 1 ) {
+				moveBlocksToPosition(
+					[ clientId ],
+					wrapperClientId,
+					getBlockRootClientId( wrapperClientId ),
+					getBlockIndex( wrapperClientId ) + 1
+				);
 				return;
 			}
 
-			event.preventDefault();
-
-			moveBlocksToPosition(
-				[ clientId ],
-				wrapperClientId,
-				getBlockRootClientId( wrapperClientId ),
-				getBlockIndex( wrapperClientId ) + 1
-			);
+			const wrapperBlock = getBlock( wrapperClientId );
+			batch( () => {
+				duplicateBlocks( [ wrapperClientId ] );
+				const blockIndex = getBlockIndex( wrapperClientId );
+				replaceInnerBlocks(
+					wrapperClientId,
+					wrapperBlock.innerBlocks.splice( 0, position )
```

Comment:
By mutating the `wrapperBlock.innerBlocks` we mess up the history that needs the specific block snapshot. We can test this easily by splitting and `undo`.

Reply:
Indeed :) I always forget that splice mutates :)